litellm/caching.py

```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+                    headers=self.headers
+                )
+            self.collection_info = collection_details.json()
+            print_verbose(f'Collection already exists.\nCollection details:{self.collection_info}')
+        else:
+            if quantization_config is None or quantization_config == 'binary':
+                quantization_params = {
+                    "binary": {
+                        "always_ram": False,
+                    }
+                }
+            elif quantization_config == 'scalar':
+                quantization_params = {
+                    "scalar": {
+                        "type": "int8",
+                        "quantile": 0.99,
+                        "always_ram": False
+                    }
+                }
+            elif quantization_config == 'product':
+                quantization_params = {
+                    "product": {
+                        "compression": "x16",
+                        "always_ram": False
+                    }
+                }
+            else: 
+                raise Exception("Quantization config must be one of 'scalar', 'binary' or 'product'")
+            
+            new_collection_status = self.sync_client.put(
+                url=f"{self.qdrant_url}/collections/{self.collection_name}",
+                json={
+                    "vectors": {
+                        "size": 1536,
+                        "distance": "Cosine"
+                    },
+                    "quantization_config": quantization_params
+                },
+                headers=self.headers
+            )
+            if new_collection_status.json()["result"]:
+                collection_details = self.sync_client.get(
+                        url=f"{self.qdrant_url}/collections/{self.collection_name}",
+                        headers=self.headers
+                    )
+                self.collection_info = collection_details.json()
+                print_verbose(f'New collection created.\nCollection details:{self.collection_info}')
+            else:
+                raise Exception("Error while creating new collection")
+
+    def _get_cache_logic(self, cached_response: Any):
+        if cached_response is None:
+            return cached_response
+        try:
+            cached_response = json.loads(
+                cached_response
+            )  # Convert string to dictionary
+        except:
+            cached_response = ast.literal_eval(cached_response)
+        return cached_response
+
+    def set_cache(self, key, value, **kwargs):
+        print_verbose(f"qdrant semantic-cache set_cache, kwargs: {kwargs}")
+        import uuid
+
+        # get the prompt
+        messages = kwargs["messages"]
+        prompt = ""
+        for message in messages:
+            prompt += message["content"]
+
+        # create an embedding for prompt
+        embedding_response = litellm.embedding(
+            model=self.embedding_model,
+            input=prompt,
```

Comment:
does the model have context window limits? I'm curious if we can allow the user to control which messages are used in the embedding request (e.g. last user message, etc.) 

Would you want that control? @haadirakhangi

Reply:
That would be great to have. However, might have to think about what is being used to create the embeddings. For instance, if we create an embedding request with last assistant response, semantic caching will most likely fail. Adding last user message makes sense though and it was something I was planning to look into later on.